docs/_posts/0000-05-01-oo-inheritance.md

```diff
@@ -23,213 +21,221 @@ Ask these questions to the group and whiteboard their answers.
-        SevenOfSpadesPlayingCard      SevenOfDiamondsPlayingCard  
-        EightOfSpadesPlayingCard      EightOfDiamondsPlayingCard  
-        NineOfSpadesPlayingCard       NineOfDiamondsPlayingCard   
-        TenOfSpadesPlayingCard        TenOfDiamondsPlayingCard    
-        JackOfSpadesPlayingCard       JackOfDiamondsPlayingCard   
-        QueenOfSpadesPlayingCard      QueenOfDiamondsPlayingCard  
-        KingOfSpadesPlayingCard       KingOfDiamondsPlayingCard   
+        ┌────────────┐                   ┌───────────┐
+        │    Card    │                   │   Deck    │
+        │            │                   │           │
+        │            │                   │ shuffle() │
+        │            │                   │           │
+        │            │                   │ deal()    │
+        └──▲─────▲───┘                   │           │
+           │     │                       │ getCards()│
+           │     │                       └──▲──────▲─┘
+           │     │                          │      │
+ ┌─────────┴─┐ ┌─┴─────────┐ ┌──────────────┴──┐ ┌─┴────────────┐
+ │PlayingCard│ │AnimalCard │ │ PlayingCardDeck │ │AnimalCardDeck│
+ │           │ │           │ │                 │ │              │
+ │snap(Card) │ │ snap(Card)│ │                 │ │              │
+ └───────────┘ └───────────┘ └─────────────────┘ └──────────────┘
 ```
 
-</span>
-
 --
 
-### Possible solutions - 20+ Classes
+## Exercise review
 
-<img src="{{ "/images/Cards3.png" | absolute_url }}" alt="Class diagram where every Suit and every FaceValue is a class in its own right">
++ How does the code compare to before?
++ What's the difference between this version and the interface based one?
+  + It's subtle and depends upon your language
+  + Inheritance is a form of abstraction 
 
 --
 
-### Composition: the other way
+## Inheritance is hard to get right
 
-+ Another way to achieve code-reuse
-+ Classes are compose of (contain instances of) other classes
-+ These instances implement the desired reusable functionality
-+ Composition avoids brittle inheritance trees
-+ Another way to achieve the same thing—composition is still OO
++ It tends to come with a rigid set of rules
++ A more common way to model relationships between objects is via _associations_
++ An _association_ is just when one class references another
 
 --
 
-### Our Vehicle example
+### Associations between our card classes
 
 ```java
-class Vehicle {
-    void accelerate(){
-        System.out.println("Getting faster");
-    }
+class PlayingCard extends Card {  //inheritance
+    private Suit suit;  //association
 }
 
-class Car extends Vehicle {
-    private SteeringWheel steeringWheel;
+class Deck {
+    protected Card[] cards;  //association
 }
 ```
 
-Note: Car contains a steering wheel  
-  Go karts and boats also require a steering wheel  
-  Car doesn’t attempt to inherit the steering wheel behaviour by inheriting it from some AbstractSteerableVehicle
-
```

Comment:
Can we keep the bit about composition please? I think it’s important to balance inheritance with the alternative. I’m happy to tidy up this a bit with some material from the design patterns book.